torch/distributed/_tools/memory_tracker.py

```diff
@@ -0,0 +1,246 @@
+from collections import defaultdict
+
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    no_type_check,
+    Sequence,
+)
+
+import torch
+import torch.nn as nn
+from torch.utils.hooks import RemovableHandle
+from torch.utils._python_dispatch import TorchDispatchMode
+
+MB = 1024 * 1024.0
+
+
+class MemoryProfileDispatchMode(TorchDispatchMode):
+    """
+    Run in TorchDispatchMode to get memory stats at operator level.
+    """
+    def __init__(self, memory_tracker) -> None:
+        self.memory_tracker = memory_tracker
+
+    def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+        rs = func(*args, **kwargs)
+        if func == torch.ops.aten.detach.default:
+            return rs
+        func_name: str = (
+            self.memory_tracker.cur_module
+            + "."
+            + func.__name__
+            + "_"
+            + str(self.memory_tracker.operator_names[func.__name__])
+        )
+        self.memory_tracker.operator_names[func.__name__] = (
+            self.memory_tracker.operator_names[func.__name__] + 1
+        )
+        self.memory_tracker._record_memory_stats(func_name)
+
+        return rs
+
+
+class MemoryTracker:
+    """
+    Collect and plot the memory stats including memories_allocated, memories_active
+    and memories_reserved at operator level.
+    It also prints a summary for the top 20 operators that generate the most memories.
+
+    Example usage:
+
+        >>> net.cuda()
+        >>> input = input.cuda()
+
+        >>> mem_tracker = MemoryTracker()
+        >>> mem_tracker.start_monitor(net)
+
+        >>> net.zero_grad(True)
+        >>> loss = net(input)
+        >>> if isinstance(loss, dict):
+        >>>    loss = loss['out']
+        >>> loss.sum().backward()
+        >>> net.zero_grad(set_to_none=True)
+
+        >>> mem_tracker.stop()
+        >>> mem_tracker.summary()
+        >>> mem_tracker.show_traces()
+    """
+
+    def __init__(self) -> None:
+        torch._C._log_api_usage_once("torch.distributed.memory_tracker")
+        self._hooks: List[RemovableHandle] = []
+        self.operator_names: Dict[str, int] = defaultdict(int)
+        self.memories_allocated: Dict[str, float] = defaultdict(float)
+        self.memories_active: Dict[str, float] = defaultdict(float)
+        self.memories_reserved: Dict[str, float] = defaultdict(float)
+        self.markers: Dict[str, int] = defaultdict(int)
+        self.cur_module: str = ""
+        self.op_id: int = 0
+
+    @no_type_check
+    def start_monitor(self, root_module: nn.Module) -> None:
+        """
+        Register module hooks and entering MemoryProfileDispatchMode, so that
+        module level and operator level memory stats can be tracked during module
+        runtime.
+        """
+        self._clear_state()
+        root_module.__setattr__("is_root", True)
```

Comment:
Should this be a public attribute `is_root`, or should we make it private?

Moreover, should we consider making an attribute that is unlikely to collide with any other use case, e.g. maybe `_memory_tracker_is_root`?